File: CapturePlugins/CSShapeCapturePlugin/CSShapeCapturePlugin/CSShapePathPlugin/CSShapePathLoader.py
import objc
import sys
import os

my_path = NSBundle.bundleForClass_(objc.lookUpClass("CSShapeCapture").class__()).resourcePath() + "/Python"
sys.path.append(my_path)

from pluginbase import PluginBase
from Foundation import *
from CSShapePathWrapper import *

sys.dont_write_bytecode = True


plugin_base = PluginBase(package='shapeplugins')

library_dirs = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSAllDomainsMask - NSSystemDomainMask, YES)
plugin_dirs = map(lambda x: x + "/Application Support/CocoaSplit/Plugins/Paths", library_dirs)
plugin_dirs.append(NSBundle.bundleForClass_(objc.lookUpClass("CSShapeCapture").class__()).resourcePath() + "/Paths")
plugin_source = plugin_base.make_plugin_source(searchpath=plugin_dirs)



class CSShapePathLoader(NSObject):
    
    def init(self):
        self = objc.super(CSShapePathLoader,self).init()
        return self
    
    @objc.signature('@@:@')
    def pathLoaderPath_(self, pluginName):
        plugin_module = plugin_source.load_plugin(pluginName)
        plugin_file = plugin_module.__file__
        real_path = os.path.realpath(plugin_file)
        return real_path
    
    
    @objc.signature('@@:')
    def allPaths(self):
        
        plugins = plugin_source.list_plugins()
        ret = {}
        for m_name in plugins:
            plugin = plugin_source.load_plugin(m_name)
            reload(plugin)
            try:
                plugin_name = plugin.name
            except AttributeError:
                plugin_name = "No Name!"
            
            new_wrap = CSShapePathWrapper.alloc().initWithPlugin_(plugin)
            
            ret[m_name] = { 'name':plugin_name, 'module':m_name, 'plugin':new_wrap}
        return ret

File: CapturePlugins/CSShapeCapturePlugin/CSShapeCapturePlugin/CSShapePathLoader.m
//
//  CSShapePathLoader.m
//  CSShapeCapturePlugin
//
//  Created by Zakk on 6/6/17.
//

#import "CSShapePathLoader.h"
#import "CSShapeWrapper.h"
#import <JavaScriptCore/JavaScriptCore.h>



@implementation CSShapePathLoader


-(void)evaluateJavascriptAtPath:(NSString *)path usingContext:(JSContext *)context
{
    
    
    NSString *scriptContents = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil];
    [context evaluateScript:scriptContents];
}



-(NSDictionary *)allPaths
{
    
    NSMutableDictionary *ret = [NSMutableDictionary dictionary];
    
    NSArray *library_dirs = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSAllDomainsMask - NSSystemDomainMask, YES);
    NSMutableArray *plugin_dirs = [NSMutableArray array];
    
    for(NSString *libDir in library_dirs)
    {
    
        [plugin_dirs addObject:[NSString stringWithFormat:@"%@/%@", libDir, @"/Application Support/CocoaSplit/Plugins/Paths"]];
    }
    NSString *bundlePath = [NSBundle bundleForClass:self.class].resourcePath;
    
    [plugin_dirs addObject:[NSString stringWithFormat:@"%@/%@", bundlePath, @"/Paths"]];
    
    
    NSPredicate *jsFilter = [NSPredicate predicateWithFormat:@"self ENDSWITH '.js'"];
    NSFileManager *fm = [NSFileManager defaultManager];
    
    
    for (NSString *pPath in plugin_dirs)
    {
        NSArray *files = [fm contentsOfDirectoryAtPath:pPath error:nil];
        NSArray *scripts = [files filteredArrayUsingPredicate:jsFilter];
        
        for (NSString *scriptPath in scripts)
        {
            JSContext *jsCtx = [[JSContext alloc] init];
            NSString *fullPath = [NSString stringWithFormat:@"%@/%@", pPath, scriptPath];
            
            [self evaluateJavascriptAtPath:fullPath usingContext:jsCtx];
            if (jsCtx[@"name"])
            {
                JSValue *jName = jsCtx[@"name"];
                NSString *name = jName.toString;
                
                CSShapeWrapper *wrapper = [[CSShapeWrapper alloc] initWithName:name usingPath:fullPath];
                ret[name] = @{@"name": name, @"path":scriptPath, @"wrapper":wrapper};
                
            }
        }
        
    }
    return ret;
}

@end
